homie-controller/tests/integration.rs

```diff
@@ -0,0 +1,226 @@
+        assert_eq!(device.homie_version, "4.0");
+        assert_eq!(device.state, State::Ready);
+        assert_eq!(device.nodes.len(), 1);
+        let node = device.nodes.get("node_id").unwrap();
+        assert_eq!(node.name, Some("Node name".to_string()));
+        assert_eq!(node.node_type, Some("node_type".to_string()));
+        assert_eq!(node.properties.len(), 1);
+        let property = node.properties.get("property_id").unwrap();
+        assert_eq!(property.name, Some("Property name".to_string()));
+        assert_eq!(property.settable, true);
+        assert_eq!(property.unit, Some("unit".to_string()));
+        assert_eq!(property.value, None);
+    }
+
+    // Send a value from the device to the controller.
+    homie
+        .publish_value("node_id", "property_id", 42)
+        .await
+        .unwrap();
+
+    // Wait until the controller receives the value.
+    loop {
+        if let Some(event) = controller.poll(&mut event_loop).await.unwrap() {
+            log::trace!("Event: {:?}", event);
+            if let Event::PropertyValueChanged {
+                device_id,
+                node_id,
+                property_id,
+                value,
+                fresh,
+            } = event
+            {
+                assert_eq!(device_id, "device_id");
+                assert_eq!(node_id, "node_id");
+                assert_eq!(property_id, "property_id");
+                assert_eq!(value, "42");
+                assert_eq!(fresh, true);
+                break;
+            }
+        }
+    }
+
+    // Check that the device looks how we expect.
+    {
+        let devices = controller.devices();
+        let device = devices.get("device_id").unwrap();
+        let node = device.nodes.get("node_id").unwrap();
+        let property = node.properties.get("property_id").unwrap();
+        log::info!("Property: {:?}", property);
+        assert_eq!(property.value(), Ok(42));
+    }
+
+    // Send a value from the controller to the device.
+    controller
+        .set("device_id", "node_id", "property_id", 13)
+        .await
+        .unwrap();
+
+    // Wait for the device to receive the value and send it back to the controller.
+    loop {
+        if let Some(event) = controller.poll(&mut event_loop).await.unwrap() {
+            log::trace!("Event: {:?}", event);
+            if let Event::PropertyValueChanged {
+                device_id,
+                node_id,
+                property_id,
+                value,
+                fresh,
+            } = event
+            {
+                assert_eq!(device_id, "device_id");
+                assert_eq!(node_id, "node_id");
+                assert_eq!(property_id, "property_id");
+                assert_eq!(value, "13");
+                assert_eq!(fresh, true);
+                break;
+            }
+        }
+    }
+    assert_eq!(
+        updates_rx.try_iter().collect::<Vec<_>>(),
+        vec!["13".to_string()]
```

Comment:
I have a vague feeling that `vec![String::from("13")] == vec!["13"]` these days (or it will be soon?)

Reply:
Aha, it is! That's nicer.